File: session/session.go
// Package session groups individual logged climbs into training sessions.
package session

import (
	"sort"
	"time"
)

type Kind int

const (
	Send Kind = iota
	Attempt
)

type Climb struct {
	Time   time.Time
	VGrade int // -1 when unknown
	Kind   Kind
	Tries  int
}

type Session struct {
	Start, End time.Time
	Climbs     []Climb
}

type Config struct {
	Gap              time.Duration
	WarmupBuffer     time.Duration
	CooldownBuffer   time.Duration
	InProgressWindow time.Duration
}

func DefaultConfig() Config {
	return Config{
		Gap:              90 * time.Minute,
		WarmupBuffer:     10 * time.Minute,
		CooldownBuffer:   5 * time.Minute,
		InProgressWindow: 2 * time.Hour,
	}
}

func Build(climbs []Climb, cfg Config, now time.Time) []Session {
	if len(climbs) == 0 {
		return nil
	}
	sorted := make([]Climb, len(climbs))
	copy(sorted, climbs)
	sort.Slice(sorted, func(i, j int) bool { return sorted[i].Time.Before(sorted[j].Time) })

	var sessions []Session
	group := []Climb{sorted[0]}
	flush := func() {
		last := group[len(group)-1]
		if now.Sub(last.Time) < cfg.InProgressWindow {
			return // still in progress; pick it up next run
		}
		sessions = append(sessions, Session{
			Start:  group[0].Time.Add(-cfg.WarmupBuffer),
			End:    last.Time.Add(cfg.CooldownBuffer),
			Climbs: group,
		})
	}
	for _, c := range sorted[1:] {
		if c.Time.Sub(group[len(group)-1].Time) > cfg.Gap {
			flush()
			group = nil
		}
		group = append(group, c)
	}
	flush()
	return sessions
}

File: packages/core/src/session.ts
export type ClimbKind = "send" | "attempt";

export type Climb = {
  time: Date;
  vGrade: number;
  name: string;
  kind: ClimbKind;
  tries: number;
  angle?: number;
};

export type Session = {
  start: Date;
  end: Date;
  climbs: Climb[];
};

export type SessionConfig = {
  gapMs: number;
  warmupBufferMs: number;
  cooldownBufferMs: number;
  inProgressWindowMs: number;
};

const MINUTE = 60_000;

export function defaultSessionConfig(): SessionConfig {
  return {
    gapMs: 90 * MINUTE,
    warmupBufferMs: 10 * MINUTE,
    cooldownBufferMs: 5 * MINUTE,
    inProgressWindowMs: 120 * MINUTE,
  };
}

export function buildSessions(climbs: Climb[], cfg: SessionConfig, now: Date): Session[] {
  if (climbs.length === 0) return [];
  const sorted = [...climbs].sort((a, b) => a.time.getTime() - b.time.getTime());

  const sessions: Session[] = [];
  let group: Climb[] = [sorted[0]!];
  const flush = (): void => {
    const last = group[group.length - 1]!;
    if (now.getTime() - last.time.getTime() < cfg.inProgressWindowMs) return;
    sessions.push({
      start: new Date(group[0]!.time.getTime() - cfg.warmupBufferMs),
      end: new Date(last.time.getTime() + cfg.cooldownBufferMs),
      climbs: group,
    });
  };
  for (const c of sorted.slice(1)) {
    if (c.time.getTime() - group[group.length - 1]!.time.getTime() > cfg.gapMs) {
      flush();
      group = [];
    }
    group.push(c);
  }
  flush();
  return sessions;
}
